doc/design_intermediate_representation.md

```diff
@@ -0,0 +1,91 @@
+# _Design:_ Intermediate Representation
+
+## Overview
+
+As SQLFlow is supporting more and more machine learning toolkits, their corresponding code generation logics are better being orgnized as separate packages. An intermediate representation(IR) of the SQL jobs becomes necessary to connect these separate packages with the core `sql` package.
+
+The core `sql` package should include the following functionalities:
+1. The entry point of running extended SQL statements.
+1. The [parsing](https://github.com/sql-machine-learning/sqlflow/blob/develop/doc/sql_parser.md) of extended SQL statements.
+1. The verification of extended SQL statements, including verifying the syntax, the existence of the selected fields.
+1. The [feature derivation](https://github.com/sql-machine-learning/sqlflow/blob/develop/doc/feature_derivation.md), including name, type, shape, and preprocessing method of the select fields.
+1. The [training data and validation data split](https://github.com/sql-machine-learning/sqlflow/blob/develop/doc/training_and_validation.md).
+
+With these functionalities, the `sql` package çan translate user typed extended SQL statements to an IR as an exposed Go struct. The codegen package takes the IR and returns a generated Python program for the `sql` package to execute.
+
+## Code Structure
+
+We propose the following code structures.
+
+```
+sql/
+  ...
+  codegen/
+    tensorflow/
+      train.go
+      predict.go
+      analyze.go
+    xgboost/
+      ...
+```
+
+The `tensorflow` package will expose function `func Train(ir sql.TrainIR) string, error`, which takes the `sql`'s `TrainIR` and returns a generated Python program.
+
+## Intermediate Representation
+
+We propose the following struct as the IR for code generation.
+
+```go
+package sql
+
+// FieldMeta contains the meta information for decoding
+type FieldMeta struct {
+	DType     string // e.g. "float", "int32"
+	Delimiter string // e.g. ","
+	Shape     []int  // e.g. [1], [1 2 3]
+	IsSparse  bool   // e.g. false
+}
+
+// TrainIR is the intermediate representation for code generation of a training job
```

Comment:
Maybe we don't need to implement an IR for each job, how about simplifying like: 
``` golang
type FeatureMeta struct {
    DType string
    Delimiter string 
    ...
}
type DBConn struct {
    Driver string
    User string
    ....
}
type ClauseIR struct {
    Estimator string
    SelectClause string
    Attributes map[string]interface{}
    DBConn DBConn
    Features map[string]FeatureMeta
    ...
}
``` 

Each generator can extend the `ClauseIR` as needed.

Reply:
@Yancey1989 Thanks for the suggestion. Combining all three `IR`s to a single `ClauseIR` does save some code. However, I still advocate using separate IRs for different job types. Here is my reasoning.
- Avoid confusion. The developer of `xgboost.Predict` would be confused by the `ValidationSelect` field in `ClauseIR`. Also, as we adding more features to SQLFlow, more fields would be added to `CluaseIR`, and the confusion will increase.
- Less work. We either distinguish the job type in `sql` or in `codegen`. However, there are many `codegen`s and only one `sql`. Distinguishing the job type in `sql` saves works in all `codegen`s.